File: src/test/java/org/springframework/samples/petclinic/e2e/pages/VisitFormPage.java
/*
 * Copyright 2012-2025 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.springframework.samples.petclinic.e2e.pages;

import com.microsoft.playwright.Locator;
import com.microsoft.playwright.Page;

/**
 * Page Object for the Visit form (add visit) page
 */
public class VisitFormPage extends BasePage {

	public VisitFormPage(Page page) {
		super(page);
	}

	public Locator getVisitDateInput() {
		return page.locator("input[name='date']");
	}

	public Locator getDescriptionInput() {
		return page.locator("input[name='description'], textarea[name='description']");
	}

	public Locator getAddVisitButton() {
		return page.getByRole(com.microsoft.playwright.options.AriaRole.BUTTON,
				new Page.GetByRoleOptions().setName("Add Visit"));
	}

	public Locator getPetInfo() {
		return page.locator("table").first();
	}

	public void fillVisitForm(String date, String description) {
		getVisitDateInput().fill(date);
		getDescriptionInput().fill(description);
	}

	public void submitForm() {
		getAddVisitButton().click();
	}

	public boolean hasValidationErrors() {
		return page.locator(".has-error, .error, .is-invalid").count() > 0;
	}

	public String getPetName() {
		return getPetInfo().locator("td").first().textContent();
	}

	public String getOwnerName() {
		return page.locator("h2").textContent();
	}

}

File: src/test/playwright/pages/VisitFormPage.js
const { BasePage } = require('./BasePage');

/**
 * VisitFormPage class for adding visits to pets
 */
class VisitFormPage extends BasePage {
  constructor(page) {
    super(page);
  }

  /**
   * Navigate to new visit form for pet
   */
  async gotoNew(ownerId, petId) {
    await super.goto(`/owners/${ownerId}/pets/${petId}/visits/new`);
  }

  /**
   * Fill visit form with data
   */
  async fillVisitForm(visitData) {
    if (visitData.date) {
      await this.page.locator('#date').fill(visitData.date);
    }
    if (visitData.description) {
      await this.page.locator('#description').fill(visitData.description);
    }
  }

  /**
   * Submit the form
   */
  async submitForm() {
    await this.page.locator('button[type="submit"]').click();
  }

  /**
   * Check if form is visible
   */
  async isFormVisible() {
    return await this.page.locator('#visit').isVisible();
  }

  /**
   * Get form title
   */
  async getFormTitle() {
    return await this.page.locator('h2').textContent();
  }

  /**
   * Get pet name displayed in form
   */
  async getPetName() {
    const petInfo = await this.page.locator('table tr:has-text("Name") td').textContent();
    return petInfo?.trim();
  }

  /**
   * Check if field has validation error
   */
  async hasFieldError(fieldName) {
    const fieldGroup = this.page.locator(`#${fieldName}`).locator('..').locator('..');
    return await fieldGroup.locator('.has-error').isVisible();
  }
}

module.exports = { VisitFormPage };